caluma/caluma_workflow/domain_logic.py

```diff
@@ -266,3 +266,32 @@ def post_cancel(case, user):
         )
 
         return case
+
+
+class CancelWorkItemLogic:
+    @staticmethod
+    def validate_for_cancel(work_item):
+        if work_item.status != models.WorkItem.STATUS_READY:
+            raise ValidationError("Only READY work items can be cancelled")
+
+    @staticmethod
+    def pre_cancel(validated_data, user):
+        validated_data["status"] = models.WorkItem.STATUS_CANCELED
+        validated_data["closed_at"] = timezone.now()
+        validated_data["closed_by_user"] = user.username
+        validated_data["closed_by_group"] = user.group
+
+        return validated_data
```

Comment:
This could be put into the `validate_for_cancel()` method. `pre_cancel()` should be used for logic that actually changes data or needs to run somewhere in the middle of the `update()`-method of the serializer.

Reply:
This uses exactly the same patterns as the other functions which were already discussed enough. Please check the context of this file and reconsider.

Reply:
Also, this obviously changes data :wink:

Reply:
As discussed: We'll leave this as it is. The `validate_for` methods are for validation only unlike the validation methods on the serializer which alter data..